### PR TITLE
allow admin change for upkeeps

### PR DESCRIPTION
**Pull request by FelixFan1992**

### CODE REVIEW

File: contracts/test/v0.8/dev/KeeperRegistryDev.test.ts [hunk 1]
```diff
@@ -527,29 +513,6 @@ describe('KeeperRegistryDev', () => {
   })
 
   describe('#unpauseUpkeep', () => {
-    context('and the registry has a paused upkeep', () => {
```
Comment (FelixFan1992) [inline contracts/test/v0.8/dev/KeeperRegistryDev.test.ts:530]: realized that this is not needed.

File: contracts/test/v0.8/dev/KeeperRegistryDev.test.ts [hunk 2]
```diff
@@ -478,20 +478,6 @@ describe('KeeperRegistryDev', () => {
   })
 
   describe('#pauseUpkeep', () => {
-    context('and the registry has an upkeep', () => {
```
Comment (FelixFan1992) [inline contracts/test/v0.8/dev/KeeperRegistryDev.test.ts:481]: not needed as we already register an upkeep before each test is run.

File: contracts/src/v0.8/dev/KeeperRegistryLogic.sol [hunk 3]
```diff
@@ -258,6 +258,36 @@ contract KeeperRegistryLogic is KeeperRegistryBase {
     emit PayeeshipTransferred(keeper, past, msg.sender);
   }
 
+  /**
+   * @dev Called through KeeperRegistry main contract
+   */
+  function transferUpkeepAdmin(uint256 id, address proposed) external {
+    Upkeep memory upkeep = s_upkeep[id];
+    if (upkeep.maxValidBlocknumber != UINT64_MAX) revert UpkeepCancelled();
+    if (upkeep.admin != msg.sender) revert OnlyCallableByAdmin();
```
Comment (infiloop2) [inline contracts/src/v0.8/dev/KeeperRegistryLogic.sol:267]: Use `requireAdminAndNotCancelled` ?

File: contracts/src/v0.8/dev/KeeperRegistryDev.sol [hunk 4]
```diff
@@ -256,6 +256,25 @@ contract KeeperRegistryDev is
     _fallback();
   }
 
+  /**
+   * @notice proposes the safe transfer of an upkeep's admin role to another address
+   * @param id the upkeep id to transfer admin
+   * @param proposed address to nominate for the new upkeep admin
+   */
+  function transferUpkeepAdmin(uint256 id, address proposed) external {
```
Comment (infiloop2) [inline contracts/src/v0.8/dev/KeeperRegistryDev.sol:264]: can you add these functions to interface?

File: contracts/src/v0.8/dev/KeeperRegistryBase.sol [hunk 5]
```diff
@@ -43,6 +43,7 @@ abstract contract KeeperRegistryBase is ConfirmedOwner, ExecutionPrevention, Ree
   mapping(uint256 => Upkeep) internal s_upkeep;
   mapping(address => KeeperInfo) internal s_keeperInfo;
   mapping(address => address) internal s_proposedPayee;
+  mapping(uint256 => address) internal s_proposedAdmin;
```
Comment (infiloop2) [inline contracts/src/v0.8/dev/KeeperRegistryBase.sol:46]: We'll also need to handle this in migration. (Can be a separate PR)

**Review by infiloop2 — CHANGES_REQUESTED** [review]
Looks good! One main comment regarding adding those functions to the interface

File: contracts/src/v0.8/dev/KeeperRegistryLogic.sol (hunk 3 above)
Comment (FelixFan1992) [reply contracts/src/v0.8/dev/KeeperRegistryLogic.sol:267]: Done

File: contracts/src/v0.8/dev/KeeperRegistryDev.sol (hunk 4 above)
Comment (FelixFan1992) [reply contracts/src/v0.8/dev/KeeperRegistryDev.sol:264]: Done

File: contracts/src/v0.8/dev/KeeperRegistryBase.sol (hunk 5 above)
Comment (FelixFan1992) [reply contracts/src/v0.8/dev/KeeperRegistryBase.sol:46]: i think this can be done by `delete s_proposedAdmin[id];` in `migrateUpkeep` ?

Comment (infiloop2) [reply contracts/src/v0.8/dev/KeeperRegistryBase.sol:46]: yes i think that will work, much simpler

**Review by infiloop2 — APPROVED** [review]
lgtm!

File: contracts/test/v0.8/dev/KeeperRegistryDev.test.ts [hunk 6]
```diff
@@ -586,20 +558,6 @@ describe('KeeperRegistryDev', () => {
   })
 
   describe('#updateCheckData', () => {
-    context('and the registry has an upkeep', () => {
```
Comment (FelixFan1992) [inline contracts/test/v0.8/dev/KeeperRegistryDev.test.ts:589]: this is not needed.

**Review by infiloop2 — APPROVED** [review]
